ComponentKit/HostingView/CKComponentHostingView.mm

```diff
@@ -40,108 +48,127 @@ - (instancetype)initWithFrame:(CGRect)frame
-    // Injected dependencies
+    _componentProvider = componentProvider;
     _sizeRangeProvider = sizeRangeProvider;
     _context = context;
-    
-    // Internal dependencies
-    _lifecycleManager = manager;
-    _lifecycleManager.delegate = self;
-    
+    _scopeRoot = [CKComponentScopeRoot rootWithListener:self];
+
     _containerView = [[CKComponentRootView alloc] initWithFrame:CGRectZero];
     [self addSubview:_containerView];
+
+    _componentNeedsUpdate = YES;
   }
   return self;
 }
 
-- (instancetype)initWithComponentProvider:(Class<CKComponentProvider>)componentProvider
-                        sizeRangeProvider:(id<CKComponentSizeRangeProviding>)sizeRangeProvider
-                                  context:(id<NSObject>)context
-{
-  CKComponentLifecycleManager *manager = [[CKComponentLifecycleManager alloc] initWithComponentProvider:componentProvider sizeRangeProvider:sizeRangeProvider];
-  return [self initWithLifecycleManager:manager sizeRangeProvider:sizeRangeProvider context:context];
-}
-
 - (void)dealloc
 {
-  [_lifecycleManager detachFromView];
+  CKAssertMainThread(); // UIKit should guarantee this
+  CKUnmountComponents(_mountedComponents);
 }
 
 #pragma mark - Layout
 
 - (void)layoutSubviews
 {
+  CKAssertMainThread();
   [super layoutSubviews];
   _containerView.frame = self.bounds;
 
-  if (_model && !CGRectIsEmpty(self.bounds)) {
-    [self _update];
-
-    if (![_lifecycleManager isAttachedToView]) {
-      [_lifecycleManager attachToView:_containerView];
+  if (!CGRectIsEmpty(self.bounds)) {
+    [self _updateComponentIfNeeded];
+    const CGSize size = self.bounds.size;
+    if (_layout.component != _component || !CGSizeEqualToSize(_layout.size, size)) {
+      _layout = [_component layoutThatFits:{size, size} parentSize:size];
     }
+    _mountedComponents = [CKMountComponentLayout(_layout, _containerView, _mountedComponents, nil) copy];
   }
 }
 
 - (CGSize)sizeThatFits:(CGSize)size
 {
-  CKSizeRange constrainedSize = [_sizeRangeProvider sizeRangeForBoundingSize:size];
-  CKComponentLayout layout = [_lifecycleManager layoutForModel:_model constrainedSize:constrainedSize context:_context];
-  return layout.size;
+  CKAssertMainThread();
+  [self _updateComponentIfNeeded];
+  const CKSizeRange constrainedSize = [_sizeRangeProvider sizeRangeForBoundingSize:size];
+  return [_component layoutThatFits:constrainedSize parentSize:constrainedSize.max].size;
 }
 
 #pragma mark - Accessors
 
 - (void)setModel:(id)model
 {
+  CKAssertMainThread();
   if (_model != model) {
     _model = model;
-    CKAssertNotNil(_model, @"Model can not be nil.");
-
-    [self setNeedsLayout];
+    [self _setComponentNeedsUpdate];
   }
 }
 
 - (void)setContext:(id<NSObject>)context
 {
+  CKAssertMainThread();
   if (_context != context) {
     _context = context;
-    [self setNeedsLayout];
+    [self _setComponentNeedsUpdate];
   }
 }
 
-- (UIView *)containerView
+- (const CKComponentLayout &)mountedLayout
```

Comment:
This is used in diamond; I made my own equivalent of CKComponentHostingView to be able to grab the layout. I do queries like hit test / bounding boxes / etc on this data structure.